sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/randomExpressions.scala

```diff
@@ -330,7 +330,8 @@ object Uniform {
   group = "string_funcs")
 case class RandStr(
     length: Expression, override val seedExpression: Expression, hideSeed: Boolean)
-  extends ExpressionWithRandomSeed with BinaryLike[Expression] with Nondeterministic {
+  extends ExpressionWithRandomSeed with BinaryLike[Expression] with Nondeterministic
+    with ExpectsInputTypes {
```

Comment:
I think we should think more about the type coercion system.

By extending `ImplicitCastInputTypes`, the function just defines the expected input data types and the type coercion system will add implicit cast following the general type coerion rule.

By extending `ExpectsInputTypes`, the function also defines the expected input data types but Spark just does type checking, no implicit CAST will be added. The function itself should implement how to get values with different input data types (can be different from CAST).

@srielau what do you think should be the future of this type coercion system?

Reply:
This is a really interesting question.
At a minimum I think any function should accept the widest possible type of the type type family it expects.
E.g. there is nothing wrong with passing a DECIMAL to a function accepting an integral numeric. Or a TINYINT where a BIGINT is expected.
The exception is if we expect that lateron we may want to overload teh function and behave different.
E.g. imagine we have ONLY +(BIGINT, BIGINT). we would NOT want DECIMAL to cast to BIGINT, because we later want +(DECIMAL, DECIMAL) etc...
Cross cvasting from e.g. a STRING to a number or a number to a STRIND is much more a matter of taste (duck-casting).